File: bench/bench_modify.jl
using JuMP
import MathOptInterface as MOI
using Gurobi
using COPT

function bench_jump(model, N::Int, M::Int)
    I = 1:N
    @variable(model, x[I])

    @constraint(model, sum(x[i] for i = I) == N)

    @objective(model, Min, sum(i / N * x[i] for i = I))

    # for j in 1:M:N
    for j in [1]
        last_variable = j + M - 1
        if last_variable > N
            last_variable = N
        end
        deleted_variables = [x[i] for i in j:last_variable]
        delete(model, deleted_variables)
        optimize!(model)
    end
end

function gurobi_model()
    optimizer = optimizer_with_attributes(
        Gurobi.Optimizer,
        "Presolve" => 0,
        "TimeLimit" => 0.0,
        MOI.Silent() => true,
    )
    model = direct_model(optimizer)
    return model
end

function copt_model()
    optimizer = optimizer_with_attributes(
        COPT.Optimizer,
        "Presolve" => 0,
        "TimeLimit" => 0.0,
        MOI.Silent() => true,
    )
    model = direct_model(optimizer)
    return model
end

function bench_gurobi(N, M)
    model = gurobi_model()
    bench_jump(model, N, M)
end

function bench_copt(N, M)
    model = copt_model()
    bench_jump(model, N, M)
end

function main(N, M)
    time_dict = Dict{String,Float64}()

    println("N: $N, M: $M")
    println("Gurobi starts")
    t1 = time()
    bench_gurobi(N, M)
    t2 = time()
    println("Gurobi ends and takes ", t2 - t1, " seconds")
    time_dict["gurobi"] = t2 - t1

    println("COPT starts")
    t1 = time()
    bench_copt(N, M)
    t2 = time()
    println("COPT ends and takes ", t2 - t1, " seconds")
    time_dict["copt"] = t2 - t1

    return time_dict
end

main(5, 1)

result_dict = Dict()
for N in [1000000]
    for M in [10000]
        result_dict[(N, M)] = main(N, M)
    end
end

result_dict

File: bench/bench_modify.py
import pyoptinterface as poi
from pyoptinterface import gurobi, copt


import gurobipy as gp
import coptpy as cp

import time


def bench_poi_base(model, N, M):
    I = range(N)
    x = poi.make_nd_variable(model, I)

    model.set_model_attribute(poi.ModelAttribute.Silent, True)
    model.set_model_attribute(poi.ModelAttribute.TimeLimitSec, 0.0)
    model.set_raw_parameter("Presolve", 0)

    expr = poi.ExprBuilder()
    for i in range(N):
        expr.add(i / N * x[i])
    model.set_objective(expr, poi.ObjectiveSense.Minimize)

    expr = poi.quicksum(x)
    model.add_linear_constraint(expr, poi.ConstraintSense.GreaterEqual, N)

    # for i in range(0, N, M):
    for i in [0]:
        last_variable = min(i + M, N)
        deleted_variables = [x[j] for j in range(i, last_variable)]
        model.delete_variables(deleted_variables)
        model.optimize()


def bench_poi_gurobi(N, M):
    model = gurobi.Model()
    bench_poi_base(model, N, M)


def bench_poi_copt(N, M):
    model = copt.Model()
    bench_poi_base(model, N, M)


def bench_gp(N, M):
    model = gp.Model()

    I = range(N)
    x = model.addVars(I)

    obj = gp.quicksum(i / N * x[i] for i in range(N))
    model.setObjective(obj)

    model.addConstr(x.sum() == N)

    model.setParam("OutputFlag", 0)
    model.setParam("TimeLimit", 0.0)
    model.setParam("Presolve", 0)

    # for j in range(0, N, M):
    for j in [0]:
        last_variable = min(j + M, N)
        for k in range(j, last_variable):
            model.remove(x[k])
        model.optimize()


def bench_cp(N, M):
    env = cp.Envr()
    model = env.createModel("m")

    I = range(N)
    x = model.addVars(I)

    model.setParam("Logging", 0)
    model.setParam("TimeLimit", 0.0)
    model.setParam("Presolve", 0)

    obj = cp.quicksum(i / N * x[i] for i in range(N))
    model.setObjective(obj)

    model.addConstr(x.sum() == N)

    # for j in range(0, N, M):
    for j in [0]:
        last_variable = min(j + M, N)
        for k in range(j, last_variable):
            x[k].remove()
        model.solve()


def main(N, M):
    result = dict()

    t1 = time.perf_counter()
    bench_poi_gurobi(N, M)
    t2 = time.perf_counter()
    result["poi_gurobi"] = t2 - t1

    t1 = time.perf_counter()
    bench_gp(N, M)
    t2 = time.perf_counter()
    result["gurobipy"] = t2 - t1

    t1 = time.perf_counter()
    bench_poi_copt(N, M)
    t2 = time.perf_counter()
    result["poi_copt"] = t2 - t1

    t1 = time.perf_counter()
    bench_cp(N, M)
    t2 = time.perf_counter()
    result["coptpy"] = t2 - t1

    return result


result_dict = dict()
for N in [1000000]:
    for M in [100000]:
        print(f"N = {N}, M = {M}")
        result = main(N, M)
        result_dict[(N, M)] = result

print(result_dict)
